optimum/habana/transformers/trainer.py

```diff
@@ -797,17 +785,47 @@ def hpu_deepspeed_checkpointing(function, *checkpoint_args):
                 if step % args.gradient_accumulation_steps == 0:
                     self.control = self.callback_handler.on_step_begin(args, self.state, self.control)
 
-                if (
-                    ((step + 1) % args.gradient_accumulation_steps != 0)
-                    and args.local_rank != -1
-                    and args._no_sync_in_gradient_accumulation
-                ):
-                    # Avoid unnecessary DDP synchronization since there will be no backward pass on this example.
-                    with model.no_sync():
-                        tr_loss_step = self.training_step(model, inputs)
+                # Proceed with forward and backward passes.
+                if args.use_hpu_graphs:
+                    # This path uses HPU graphs, where the entire FW+BWD computation graph is scheduled on Gaudi within milliseconds of CPU time.
+                    if hasattr(self, "_recorded_fwbwd"):
+                        fwbwd_graph, fwbwd_inputs, fwbwd_output = self._recorded_fwbwd
+                    else:
+                        fwbwd_inputs = {}
+                        for input_key, input_value in inputs.items():
+                            fwbwd_inputs[input_key] = torch.zeros(
+                                size=input_value.shape, dtype=input_value.dtype, device="hpu:0"
+                            )
+
+                        import habana_frameworks.torch as ht
+
+                        fwbwd_graph = ht.hpu.HPUGraph()
+                        with ht.hpu.stream(ht.hpu.Stream()):
+                            fwbwd_graph.capture_begin()
+                            fwbwd_output = self.training_step(model, fwbwd_inputs)
+                            fwbwd_graph.capture_end()
+
+                        self._recorded_fwbwd = (fwbwd_graph, fwbwd_inputs, fwbwd_output)
+
+                    for input_key, input_value in inputs.items():
+                        fwbwd_inputs[input_key].copy_(input_value, non_blocking=True)
+
+                    fwbwd_graph.replay()
+                    tr_loss_step = fwbwd_output
                 else:
                     tr_loss_step = self.training_step(model, inputs)
 
+                is_optimization_step = (step + 1) % args.gradient_accumulation_steps == 0 or (
+                    # last step in epoch but step is always smaller than gradient_accumulation_steps
+                    steps_in_epoch <= args.gradient_accumulation_steps
+                    and (step + 1) == steps_in_epoch
+                )
+
+                if self.args.world_size > 1 and is_optimization_step:
+                    all_reduce_gradients(
+                        model, use_hpu_graph=True
```

Comment:
Not using HPU graphs here works slower than original DDP path (there are almost 800 memcopies issued under the hood, which costs CPU time).